src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java

```diff
@@ -442,6 +442,26 @@ Set<String> verifyInputColumns(
               "Values for all non-nullable columns must be specified, rowIndex:%d", rowIndex));
     }
 
+    // Check for null values in not-nullable columns in the row
+    List<String> nullValueNotNullCols = new ArrayList<>();
+    for (String columnName : this.nonNullableFieldNames) {
+      if (inputColNamesMap.containsKey(columnName)
+          && row.get(inputColNamesMap.get(columnName)) == null) {
+        nullValueNotNullCols.add(statsMap.get(columnName).getColumnDisplayName());
+      }
+    }
```

Comment:
We already throw an exception at the end of `parseColumnValueToParquet` when we insert null into a non-nullable column, should be just catch that and add the col name to the nullValueNotNullCols list?

Reply:
I'm not sure if it is the right place. The `verifyInputColumns` is called before and sounds like the right place for such validation. `verifyInputColumns` is also in the abstract class, the `parseColumnValueToParquet` only in one implementation (Parquet).

If we catched the exception in insertRows function we would only add single column to the nullValueNotNullCols even if there is more columns with null values. We would need to pass the error object into the `addRow` function to catch all the exception that may be thrown in a loop.

Therefore i think the `verifyInputColumns` is much better place for the validation

Reply:
I see, makes sense to me, my concern was that in the worst case, we're iterating through all the data again to check this which is not ideal so I was hoping that we can do this as part of data validation.